### PR TITLE
Added 'meta_keywords' to search of products

### LINKED ISSUE #1936: Using Meta Keywords to Improve Search
The Problem: Currently the Product search box allows searching in the Category, Product Title and Variant Name. This is important as descriptions often contain a lot of additional text we don't want to search for. However, this means that products can end up with long names if we want to be able to search for them but don't want to create a specific category for the product. An example from the UK is craft items like Greeting Cards.

Proposed Solution: To use the SEO fields in the Edit Product screen to do this. I would propose that we use the 'Meta Keyword' field. Thus terms that appear in the 'Meta Keyword' field will be searchable in the Product Search and thus be displayed to the user at that point.

Some fields will need to be renamed for UX:
Tab Name SEO could become 'Search'
'Meta Keyword' could become 'Product Search Keywords'
'Meta Description' could become 'SEO Keywords'.

Fixing the SEO keywords is outside the scope of this issue.
It would be useful to have a tooltip to explain what this is doing.

image

Discussion: https://community.openfoodnetwork.org/t/enabling-meta-keywords-to-help-with-product-search/1102

### PR DESCRIPTION
**Pull request by Duende13**
Changed labels on SEO tab and SEO form (edit product Screen) and added in en.yml file.

#### What? Why?

Closes #1936 

It was proposed as solution to be able to look for 'meta_keywords' to avoid that users have to add long names to describe their products and variants, they can add more details into the meta_keywords field, renamed  'Product Search Keywords' to achieve that.

#### What should we test?

To search for products where the meta_keywords or new  'Product Search Keywords' field has been used and applies the filter to find them.

### CODE REVIEW

File: app/views/spree/admin/products/seo.html.haml [hunk 1]
```diff
@@ -1,5 +1,5 @@
 = render :partial => 'spree/admin/shared/product_sub_menu'
-= render :partial => 'spree/admin/shared/product_tabs', :locals => { :current => 'SEO' }
+= render :partial => 'spree/admin/shared/product_tabs', :locals => { :current => 'Search' }
```
Comment (sauloperez) [inline app/views/spree/admin/products/seo.html.haml]: This should be translated instead of hardcoded

File: spec/features/admin/products_spec.rb [hunk 2]
```diff
@@ -184,19 +184,17 @@
...
+      within('#sidebar') { click_link 'Search' }
+      fill_in 'Product Search Keywords', :with => 'Product Search Keywords'
       fill_in 'Notes', :with => 'Just testing Notes'
       click_button 'Update'
       flash_message.should == "Product \"#{product.name}\" has been successfully updated!"
       product.reload
       product.notes.should == 'Just testing Notes'
-      product.meta_keywords.should == 'Meta Keywords'
-      product.meta_description.should == 'Meta Description'
+      product.meta_keywords.should == 'Product Search Keywords'
```
Comment (sauloperez) [inline spec/features/admin/products_spec.rb]: please, use RSpec's `expect` syntax :pray:

File: spec/features/consumer/shopping/shopping_spec.rb [hunk 3]
```diff
@@ -212,6 +212,10 @@
         fill_in "search", with: "Meer"           # For product named "Meercats"
         page.should have_content product2.name
         page.should_not have_content product.name
+
+        fill_in "search", with: "Dome"           # For product with meta_keywords "Domestic"
+        page.should have_content product.name
+        page.should_not have_content product2.name
```
Comment (sauloperez) [inline spec/features/consumer/shopping/shopping_spec.rb]: here as well

File: app/views/spree/admin/products/_seo_form.html.haml [hunk 4]
```diff
@@ -1,11 +1,11 @@
 .row{"data-hook" => "admin_product_meta_form"}
   .alpha.eleven.columns
     = f.field_container :meta_description do
-      = f.label :meta_keywords, t(:meta_keywords)
+      = f.label :meta_keywords, t(:product_search_keywords)
       %br/
       = f.text_field :meta_keywords, :class => 'fullwidth', :rows => 6
     = f.field_container :meta_description do
-      = f.label :meta_description, t(:meta_description)
```
Comment (sauloperez) [inline app/views/spree/admin/products/_seo_form.html.haml]: Is there a reason why we can't simply replace the translations of these two terms?

**Comment by Duende13** [comment]
Hi @sauloperez, I just made the changes requested, I hope everything is ok now!

File: app/views/spree/admin/products/_seo_form.html.haml (hunk 4 above)
Comment (Duende13) [reply app/views/spree/admin/products/_seo_form.html.haml]: Hi @sauloperez , I thought that you had the wrong file, because in the one I changed the meta_description label has already changed to 'SEO_keywords', which I added on the  en.yml file to be translated, not sure now if you meant another thing?

Comment (sauloperez) [reply app/views/spree/admin/products/_seo_form.html.haml]: > @lin-d-hop and I agreed to change the labels to add more readability to the code

In that case, can you check whether the previous keys are used at all and can be removed?

Comment (Duende13) [reply app/views/spree/admin/products/_seo_form.html.haml]: For this reason we changed them, because there were not keys yet introduced on the en.yml other any other local yml file. I couldn't find these keys in the whole project, just in the form ;) (what actually let me a little bit surprised )

Comment (sauloperez) [reply app/views/spree/admin/products/_seo_form.html.haml]: aha! good catch!

**Comment by daniellemoorhead** [comment]
@lin-d-hop @sineadfenton will leave it in your capable hands to get this staged in the UK and tested. Let us know if it's curly enough functionality to need Sally to test it too. Cheers!

**Comment by daniellemoorhead** [comment]
@lin-d-hop @sineadfenton are we ready to get back onto this PR and get it tested and merged? Perhaps you need some testing help from @myriamboure who is onboard and able to test?

**Comment by lin-d-hop** [comment]
Hi @Duende13 
Could you help me out as to where I might be going wrong here?
1) I search the products using a keyword 'banana' BEFORE I add it:
screenshot from 2018-02-09 13-13-23
No result = AS EXPECTED

2) I add the keyword 'banana' to the list of keywords for a product on this supplier
screenshot from 2018-02-09 13-14-05

3) I search the products again using the keyword 'banana' AFTER adding it.
screenshot from 2018-02-09 13-27-45
No result = UNEXPECTED BEHAVIOUR

Did I follow the correct steps?
Thanks!

**Comment by mkllnk** [comment]
Staged on https://staging1.openfood.com.au/. @myriamboure

**Comment by HugsDaniel** [comment]
I added the tooltips but I can't manage to make a pull request keeping @Duende13 commits... Could I make a new pull request with just the tooltips and we keep this PR as it is ? It's actually not a big modification.

**Comment by myriamboure** [comment]
Probably @enricostano or @sauloperez can give you some advices :-)

**Comment by enricostano** [comment]
@HugsDaniel to keep her commits you should:
- fork her fork (:sweat_smile: )
- add your commits to her branch
- open a PR from your fork to upstream
- close this PR and put a link it in your PR

Makes sense?

**Comment by HugsDaniel** [comment]
I can't fork her fork and I can't fork openfoodnetwork twice either :/ I was able to push my commits and have hers, but I can't make a PR to openfoodnetwork master branch...

**Comment by sauloperez** [comment]
>  I was able to push my commits and have hers

where? in a fork of yours?

**Comment by HugsDaniel** [comment]
I cloned her repo, created a new one on my github and pushed the clone to this new repo I created. But it's not a fork so Github doesn't link it to OFN main repo.

**Comment by sauloperez** [comment]
I'd rather fork from OFN, and apply these commits you already did (including her's) by means of a `git patch` (it should work :trollface: ).

**Comment by enricostano** [comment]
Ouch, well... another way to keep her commits would be to get in touch with her and ask permissions to pussh to her fork.

**Comment by sauloperez** [comment]
Oh yes, I forgot humans can talk to each other. We forget it too many times :trollface: :see_no_evil:

**Comment by HugsDaniel** [comment]
permission
@myriamboure

**Comment by myriamboure** [comment]
@sauloperez @enricostano we found the issue with @HugsDaniel it's that he doesn't have "write" permission on the OFN repo. I understand we do manage carefully those rights but we should agree on clear rules and maybe have some short term exception where we can give write rights temporarly (in those cases for instance)? Anyway I saw that the team "UK devs and maintainers" and "Devs&review" have write permission on Github for now, maybe we need just one team that has write permission to avoid loosing track and define clearly the boundaries? Like should it be only the few core devs involved daily? Or other more punctual devs (like in the UK repo)? 
I don't have the write on teams and permission so couldn't add Hugo anywhere, we agreed that he would copy paste Suzanna work in a brand new PR to avoid losing time again as he wanted to finish that tonight. He will link his PR to this one to acknowledge Suzanna work :-)

**Comment by HugsDaniel** [comment]
Ok so I just took @Duende13 's modifications and re-made the commits. It's under my name, but as I say in #2134, all credits go to @Duende13. Is it ok for you guys ?

**Comment by mkllnk** [comment]
@myriamboure He would have only needed OFN write permissions to change the code in Susana's repository. While we do that regularly, it doesn't mean it's the only way or the best practice. 

@HugsDaniel Great that you found a solution by re-creating those commits. Did you know that you can actually change the author as well? For example: `git commit --amend --author="other.person@example.com" --date "2005-04-07T22:13:13"`

But in this case, you didn't need to re-create them. You can add Susana's repository as remote and then create a new branch locally:
```sh
git remote add duende13 git@github.com:Duende13/openfoodnetwork.git
git fetch duende13
git checkout duende13/su_issue_1936 -b su_issue_1936
# do your modifications like:
git cherry-pick hugs-new-commit
git push origin
# or `git push hugsdaniel` if that's your remote
```

For now, I will add your commit to this pull request to keep Susana as author.

File: app/views/spree/admin/products/seo.html.haml [hunk 5]
```diff
@@ -2,7 +2,8 @@
 = render :partial => 'spree/admin/shared/product_tabs', :locals => { :current => t(:Search) }
 = render :partial => 'spree/shared/error_messages', :locals => { :target => @product }
 
-= form_for [:admin, @product], :method => :put, :html => { :multipart => true } do |f|
-  %fieldset.no-border-top
-    = render :partial => 'seo_form', :locals => { :f => f }
-    = render :partial => 'spree/admin/shared/edit_resource_links'
+%div{ 'ng-app' => 'ofn.admin' }
```
Comment (mkllnk) [inline app/views/spree/admin/products/seo.html.haml]: Why did you add the app? What that necessary to use `ofn-with-tip`?

**Comment by daniellemoorhead** [comment]
giphy

😆

**Comment by sigmundpetersen** [comment]
May the forks be with you

File: app/views/spree/admin/products/seo.html.haml (hunk 5 above)
Comment (HugsDaniel) [reply app/views/spree/admin/products/seo.html.haml]: I'm no Angular specialist but it seems to be the way to link this view to the controller so that the tip is triggered.

**Comment by myriamboure** [comment]
@mkllnk are you happy with @HugsDaniel answer? Enrico and Pau moved to test ready but wanted to make sure you agree :-)

**Comment by myriamboure** [comment]
@HugsDaniel @mkllnk that branch has conflicts that must be resolved, we can't test :-( Can one of you rebase? Hope you will have the rights to do that Hugo :-) Thanks!

**Comment by lin-d-hop** [comment]
or @Matt-Yorkley?

**Comment by sauloperez** [comment]
solved!

**Comment by myriamboure** [comment]
Deployed on https://staging.openfoodnetwork.org.uk/

**Comment by myriamboure** [comment]
It seems that something strange is happening here. Tooltips are fixed, but the basic feature is not working anymore. What triggers some red light in my head is that @lin-d-hop had depoyed that previously and it was not working but when @mkllnk deployed in Australia it was working (with same code, Lynne was actually surprised). So may I suggest we deploy that branch in another staging like Katuma @sauloperez if you have the time to stage that today? Else I'll ask Maikel on wednesday. Thanks!

**Comment by myriamboure** [comment]
wait! It seems it might just be a very longgggg delay in UK staging taking that into account wait a minute I'll put updates, I got something working.

**Comment by myriamboure** [comment]
Ok this is very strange... I don't know what happened when tooltips were added @mkllnk @HugsDaniel but it broke something / or there is an issue with UK staging.
Before when staged in Aus: I could type 2 keywords separate with space and search was working on both.
Now after fix on tooltips and staged in UK: If I type one keyword and search in shop it works, but if I add a second keyword it stops working AND if I add a keyword and then delete it so only one remaining it's not working back again... see the end of testing notes: https://docs.google.com/document/d/16F_C9e4pfi0FHpwgHkMbb4f15aFqauOeay2NRVEvzCA/edit#

Not sure what to do... shall we try to deploy on another staging as might be something UK staging related? (cf previous issues we had) What do you think, could the addition from Hugo and Maikel (who reported Hugo's changes here) have broken something?

**Comment by mkllnk** [comment]
Staged on https://staging1.openfood.com.au/. @myriamboure

**Comment by myriamboure** [comment]
OK so that's what I thought, it  works perfectly in Aus staging, see end of testing notes: https://docs.google.com/document/d/16F_C9e4pfi0FHpwgHkMbb4f15aFqauOeay2NRVEvzCA/edit?usp=sharing
So how to move forward with that? How can it be that it works on Aus staging but not in UK staging? I'm a bit afraid to merge this if we don't understand why that happens... if I would have tested ony in Aus staging I would have directly send to ready to go... any idea @mkllnk @HugsDaniel @sauloperez @enricostano ?

**Comment by sauloperez** [comment]
could it simply be that this branch wasn't actually deployed?

**Comment by myriamboure** [comment]
No it can't I clearly saw the tooltips on UK staging and also it was working when only one keyword ...

**Comment by sauloperez** [comment]
Staged on https://staging.katuma.org/

**Comment by myriamboure** [comment]
Ok so works perfectly on Katuma staging (see end of testing notes: https://docs.google.com/document/d/16F_C9e4pfi0FHpwgHkMbb4f15aFqauOeay2NRVEvzCA/edit?usp=sharing) SO there is a problem with UK staging @lin-d-hop I don't know what it is though... how can a deployment issue make his feature work or not work? That's a mistery for me, but I would say this is ready to go.

**Comment by myriamboure** [comment]
Ok so I retested on UK staging just now and redid exactly the same tests as yesterday and now it is working... I don't understand, that's mysterious to me. I might have done a mistake but when I read again the testing notes I documented all what I did yesterday on UK staging and I know I tested many times and cases because I couldn't understand why and when it was working or not. SO let's say maybe it was a temporary problem on UK staging, maybe it was super slow in analyzing the keywords and I was not patient enough for the results to appear, I don't know. So let's merge that now that it works basically on all stagings ;-) @lin-d-hop just be aware about that "thing" that happened on UK staging maybe for cases to come...
